File: packages/typescript/src/orchestrator.ts
import type {
  AgentRole,
  HandoffRequest,
  OrchestratorConfig,
  Pipeline,
  PipelineStage,
  SharedContext,
  Task,
  TaskResult,
  TaskStatus,
  TaskType,
} from "./types.js";
import { AgentRole as Roles, TaskStatus as Status, TaskType as Types } from "./types.js";
import type { Agent } from "./agent.js";

const DEFAULT_PIPELINES: Pipeline[] = [
  {
    name: "Backend Feature",
    taskType: Types.BackendFeature,
    stages: [
      { role: Roles.TechLead, action: "plan", required: true },
      { role: Roles.BackendDeveloper, action: "implement", required: true },
      { role: Roles.QAEngineer, action: "test", required: true },
    ],
  },
  {
    name: "Frontend Feature",
    taskType: Types.FrontendFeature,
    stages: [
      { role: Roles.TechLead, action: "plan", required: true },
      { role: Roles.FrontendDeveloper, action: "implement", required: true },
      { role: Roles.QAEngineer, action: "test", required: true },
    ],
  },
  {
    name: "Bug Fix",
    taskType: Types.BugFix,
    stages: [
      { role: Roles.QAEngineer, action: "reproduce", required: true },
      { role: Roles.BackendDeveloper, action: "fix", required: true },
      { role: Roles.QAEngineer, action: "verify", required: true },
    ],
  },
  {
    name: "Database Migration",
    taskType: Types.DBMigration,
    stages: [
      { role: Roles.DBA, action: "design", required: true },
      { role: Roles.BackendDeveloper, action: "integrate", required: true },
      { role: Roles.QAEngineer, action: "verify", required: true },
    ],
  },
  {
    name: "Security Review",
    taskType: Types.SecurityReview,
    stages: [
      { role: Roles.SecurityAnalyst, action: "audit", required: true },
      { role: Roles.BackendDeveloper, action: "fix", required: false },
      { role: Roles.SecurityAnalyst, action: "verify", required: true },
    ],
  },
  {
    name: "Documentation",
    taskType: Types.Documentation,
    stages: [
      { role: Roles.TechnicalWriter, action: "write", required: true },
      { role: Roles.TechLead, action: "review", required: false },
    ],
  },
  {
    name: "Infrastructure",
    taskType: Types.Infrastructure,
    stages: [
      { role: Roles.DevOpsEngineer, action: "implement", required: true },
      { role: Roles.SecurityAnalyst, action: "review", required: false },
    ],
  },
  {
    name: "Testing",
    taskType: Types.Testing,
    stages: [
      { role: Roles.QAEngineer, action: "write-tests", required: true },
    ],
  },
  {
    name: "Code Review",
    taskType: Types.CodeReview,
    stages: [
      { role: Roles.TechLead, action: "review", required: true },
      { role: Roles.SecurityAnalyst, action: "security-check", required: false },
    ],
  },
  {
    name: "Sprint Planning",
    taskType: Types.SprintPlanning,
    stages: [
      { role: Roles.ProjectManager, action: "plan", required: true },
      { role: Roles.TechLead, action: "estimate", required: true },
    ],
  },
];

export class Orchestrator {
  private agents: Map<AgentRole, Agent> = new Map();
  private pipelines: Pipeline[];
  private config: OrchestratorConfig;

  constructor(config: OrchestratorConfig = {}) {
    this.config = config;
    this.pipelines = config.pipelines ?? DEFAULT_PIPELINES;
  }

  /**
   * Register an agent with the orchestrator.
   */
  registerAgent(agent: Agent): void {
    this.agents.set(agent.role, agent);
  }

  /**
   * Register multiple agents at once.
   */
  registerAgents(agents: Agent[]): void {
    for (const agent of agents) {
      this.registerAgent(agent);
    }
  }

  /**
   * Get all registered agent roles.
   */
  getRegisteredRoles(): AgentRole[] {
    return Array.from(this.agents.keys());
  }

  /**
   * Get the pipeline for a given task type.
   */
  getPipeline(taskType: TaskType): Pipeline | undefined {
    return this.pipelines.find((p) => p.taskType === taskType);
  }

  /**
   * Submit a task for processing through its pipeline.
   * Returns results from each stage in order.
   */
  async submitTask(task: Task): Promise<TaskResult[]> {
    const pipeline = this.getPipeline(task.type);
    if (!pipeline) {
      throw new Error(`No pipeline defined for task type: ${task.type}`);
    }

    const context: SharedContext = {
      taskId: task.id,
      history: [],
      files: {},
      metadata: {},
    };

    const results: TaskResult[] = [];

    for (const stage of pipeline.stages) {
      const agent = this.agents.get(stage.role);

      if (!agent) {
        if (stage.required) {
          throw new Error(
            `Required agent role "${stage.role}" is not registered. ` +
            `Pipeline "${pipeline.name}" cannot proceed.`
          );
        }
        continue;
      }

      task.status = Status.InProgress as TaskStatus;
      task.assignedTo = stage.role;

      try {
        const result = await agent.execute(task, context);
        results.push(result);

        this.config.onTaskComplete?.(result);

        if (result.handoffRequest) {
          this.config.onHandoff?.(result.handoffRequest);
          const handoffResult = await this.handleHandoff(
            result.handoffRequest,
            task,
            context
          );
          if (handoffResult) {
            results.push(handoffResult);
          }
        }

        if (!result.success && stage.required) {
          task.status = Status.Failed as TaskStatus;
          return results;
        }
      } catch (error) {
        const err = error instanceof Error ? error : new Error(String(error));
        this.config.onError?.(err, task);

        if (stage.required) {
          task.status = Status.Failed as TaskStatus;
          throw err;
        }
      }
    }

    task.status = Status.Completed as TaskStatus;
    return results;
  }

  /**
   * Process a handoff request by routing to the target agent.
   */
  private async handleHandoff(
    handoff: HandoffRequest,
    task: Task,
    context: SharedContext
  ): Promise<TaskResult | null> {
    const agent = this.agents.get(handoff.toRole);
    if (!agent) {
      return null;
    }

    context.metadata.handoffReason = handoff.reason;
    context.metadata.handoffFrom = handoff.fromRole;
    context.metadata.handoffContext = handoff.context;

    return agent.execute(task, context);
  }
}

File: packages/python/src/tcs_bpi_ai_team/orchestrator.py
"""Orchestrator — deterministic task routing through agent pipelines."""

from __future__ import annotations

from tcs_bpi_ai_team.agent import Agent
from tcs_bpi_ai_team.types import (
    AgentRole,
    HandoffRequest,
    OrchestratorConfig,
    Pipeline,
    PipelineStage,
    SharedContext,
    Task,
    TaskResult,
    TaskStatus,
    TaskType,
)

R = AgentRole
T = TaskType

DEFAULT_PIPELINES: list[Pipeline] = [
    Pipeline(
        name="Backend Feature",
        task_type=T.BACKEND_FEATURE,
        stages=[
            PipelineStage(role=R.TECH_LEAD, action="plan", required=True),
            PipelineStage(role=R.BACKEND_DEVELOPER, action="implement", required=True),
            PipelineStage(role=R.QA_ENGINEER, action="test", required=True),
        ],
    ),
    Pipeline(
        name="Frontend Feature",
        task_type=T.FRONTEND_FEATURE,
        stages=[
            PipelineStage(role=R.TECH_LEAD, action="plan", required=True),
            PipelineStage(role=R.FRONTEND_DEVELOPER, action="implement", required=True),
            PipelineStage(role=R.QA_ENGINEER, action="test", required=True),
        ],
    ),
    Pipeline(
        name="Bug Fix",
        task_type=T.BUG_FIX,
        stages=[
            PipelineStage(role=R.QA_ENGINEER, action="reproduce", required=True),
            PipelineStage(role=R.BACKEND_DEVELOPER, action="fix", required=True),
            PipelineStage(role=R.QA_ENGINEER, action="verify", required=True),
        ],
    ),
    Pipeline(
        name="Database Migration",
        task_type=T.DB_MIGRATION,
        stages=[
            PipelineStage(role=R.DBA, action="design", required=True),
            PipelineStage(role=R.BACKEND_DEVELOPER, action="integrate", required=True),
            PipelineStage(role=R.QA_ENGINEER, action="verify", required=True),
        ],
    ),
    Pipeline(
        name="Security Review",
        task_type=T.SECURITY_REVIEW,
        stages=[
            PipelineStage(role=R.SECURITY_ANALYST, action="audit", required=True),
            PipelineStage(role=R.BACKEND_DEVELOPER, action="fix", required=False),
            PipelineStage(role=R.SECURITY_ANALYST, action="verify", required=True),
        ],
    ),
    Pipeline(
        name="Documentation",
        task_type=T.DOCUMENTATION,
        stages=[
            PipelineStage(role=R.TECHNICAL_WRITER, action="write", required=True),
            PipelineStage(role=R.TECH_LEAD, action="review", required=False),
        ],
    ),
    Pipeline(
        name="Infrastructure",
        task_type=T.INFRASTRUCTURE,
        stages=[
            PipelineStage(role=R.DEVOPS_ENGINEER, action="implement", required=True),
            PipelineStage(role=R.SECURITY_ANALYST, action="review", required=False),
        ],
    ),
    Pipeline(
        name="Testing",
        task_type=T.TESTING,
        stages=[
            PipelineStage(role=R.QA_ENGINEER, action="write-tests", required=True),
        ],
    ),
    Pipeline(
        name="Code Review",
        task_type=T.CODE_REVIEW,
        stages=[
            PipelineStage(role=R.TECH_LEAD, action="review", required=True),
            PipelineStage(role=R.SECURITY_ANALYST, action="security-check", required=False),
        ],
    ),
    Pipeline(
        name="Sprint Planning",
        task_type=T.SPRINT_PLANNING,
        stages=[
            PipelineStage(role=R.PROJECT_MANAGER, action="plan", required=True),
            PipelineStage(role=R.TECH_LEAD, action="estimate", required=True),
        ],
    ),
]


class Orchestrator:
    """Deterministic task orchestrator that routes tasks through agent pipelines."""

    def __init__(self, config: OrchestratorConfig | None = None) -> None:
        self.config = config or OrchestratorConfig()
        self._agents: dict[AgentRole, Agent] = {}
        self._pipelines = (
            list(DEFAULT_PIPELINES)
            if self.config.pipelines is None
            else self.config.pipelines
        )

    def register_agent(self, agent: Agent) -> None:
        """Register an agent with the orchestrator."""
        self._agents[agent.role] = agent

    def register_agents(self, agents: list[Agent]) -> None:
        """Register multiple agents at once."""
        for agent in agents:
            self.register_agent(agent)

    def get_registered_roles(self) -> list[AgentRole]:
        """Get all registered agent roles."""
        return list(self._agents.keys())

    def get_pipeline(self, task_type: TaskType) -> Pipeline | None:
        """Get the pipeline for a given task type."""
        for pipeline in self._pipelines:
            if pipeline.task_type == task_type:
                return pipeline
        return None

    async def submit_task(self, task: Task) -> list[TaskResult]:
        """Submit a task for processing through its pipeline.

        Returns results from each stage in order.
        """
        pipeline = self.get_pipeline(task.type)
        if pipeline is None:
            msg = f"No pipeline defined for task type: {task.type}"
            raise ValueError(msg)

        context = SharedContext(task_id=task.id)
        results: list[TaskResult] = []

        for stage in pipeline.stages:
            agent = self._agents.get(stage.role)

            if agent is None:
                if stage.required:
                    msg = (
                        f'Required agent role "{stage.role}" is not registered. '
                        f'Pipeline "{pipeline.name}" cannot proceed.'
                    )
                    raise ValueError(msg)
                continue

            task.status = TaskStatus.IN_PROGRESS
            task.assigned_to = stage.role

            try:
                result = await agent.execute(task, context)
                results.append(result)

                if self.config.on_task_complete:
                    self.config.on_task_complete(result)

                if result.handoff_request:
                    if self.config.on_handoff:
                        self.config.on_handoff(result.handoff_request)
                    handoff_result = await self._handle_handoff(
                        result.handoff_request, task, context
                    )
                    if handoff_result:
                        results.append(handoff_result)

                if not result.success and stage.required:
                    task.status = TaskStatus.FAILED
                    return results

            except Exception as err:
                if self.config.on_error:
                    self.config.on_error(err, task)

                if stage.required:
                    task.status = TaskStatus.FAILED
                    raise

        task.status = TaskStatus.COMPLETED
        return results

    async def _handle_handoff(
        self,
        handoff: HandoffRequest,
        task: Task,
        context: SharedContext,
    ) -> TaskResult | None:
        """Process a handoff request by routing to the target agent."""
        agent = self._agents.get(handoff.to_role)
        if agent is None:
            return None

        context.metadata["handoff_reason"] = handoff.reason
        context.metadata["handoff_from"] = handoff.from_role
        context.metadata["handoff_context"] = handoff.context

        return await agent.execute(task, context)
